File: bin/harness.rs
use std::path::{Path, PathBuf};
use std::time::Duration;

pub(super) const SOURCE_AGENT_FILE: &str = "AGENTS.md";
const CLAUDE_AGENT_FILE: &str = "CLAUDE.md";
const INSTALL_TIMEOUT_SECONDS: u64 = 120;
pub(super) const MANAGED_STATE_SUBDIR: &str = ".local/share/agents/sync-managed";

pub(super) type AssetRename = (&'static str, &'static str);

const PI_COMPAT_MANAGED_ENTRIES: &[&str] = &["legacy"];

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub(super) enum HarnessId {
    Claude,
    Codex,
    Opencode,
    Pi,
    Omp,
}

#[derive(Clone, Debug)]
pub(super) struct Harness {
    pub(super) id: HarnessId,
    pub(super) source_name: &'static str,
    home: PathBuf,
    instruction_file: &'static str,
    asset_renames: &'static [AssetRename],
    runtime_subdir: Option<&'static str>,
    pub(super) compat_managed_entries: &'static [&'static str],
}

impl Harness {
    fn new(id: HarnessId, source_name: &'static str, home: PathBuf) -> Self {
        Self {
            id,
            source_name,
            home,
            instruction_file: SOURCE_AGENT_FILE,
            asset_renames: &[],
            runtime_subdir: None,
            compat_managed_entries: &[],
        }
    }

    fn with_instruction_file(mut self, instruction_file: &'static str) -> Self {
        self.instruction_file = instruction_file;
        self
    }

    fn with_runtime_subdir(mut self, runtime_subdir: &'static str) -> Self {
        self.runtime_subdir = Some(runtime_subdir);
        self
    }

    fn with_compat_managed_entries(
        mut self,
        compat_managed_entries: &'static [&'static str],
    ) -> Self {
        self.compat_managed_entries = compat_managed_entries;
        self
    }

    pub(super) fn root(&self) -> PathBuf {
        self.runtime_subdir
            .map_or_else(|| self.home.clone(), |subdir| self.home.join(subdir))
    }

    pub(super) fn source_root(&self, tools_home: &Path) -> PathBuf {
        self.runtime_subdir.map_or_else(
            || tools_home.join(self.source_name),
            |subdir| tools_home.join(self.source_name).join(subdir),
        )
    }

    pub(super) fn instruction_file_name(&self) -> &'static str {
        self.instruction_file
    }

    pub(super) fn instruction_target(&self) -> PathBuf {
        self.root().join(self.instruction_file)
    }

    pub(super) fn rename_asset(&self, asset_name: &str) -> String {
        self.asset_renames
            .iter()
            .find_map(|(src, dst)| (*src == asset_name).then_some((*dst).to_string()))
            .unwrap_or_else(|| asset_name.to_string())
    }

    pub(super) fn managed_state_path(&self, managed_state_home: &Path) -> PathBuf {
        managed_state_home.join(format!("{}.json", self.source_name))
    }
}

#[derive(Clone, Debug)]
pub(super) struct SyncEnv {
    pub(super) home: PathBuf,
    pub(super) assets_home: PathBuf,
    pub(super) tools_home: PathBuf,
    pub(super) mcporter_home: PathBuf,
    pub(super) managed_state_home: PathBuf,
    pub(super) install_timeout: Duration,
    pub(super) harnesses: Vec<Harness>,
}

impl SyncEnv {
    pub(super) fn from_system() -> Result<Self, String> {
        let home = std::env::var_os("HOME")
            .map(PathBuf::from)
            .ok_or_else(|| "missing HOME".to_string())?;
        Ok(Self::from_home(
            home,
            Duration::from_secs(INSTALL_TIMEOUT_SECONDS),
        ))
    }

    pub(super) fn from_home(home: PathBuf, install_timeout: Duration) -> Self {
        let agents_home = home.join(".config").join("agents");
        Self {
            home: home.clone(),
            assets_home: agents_home.join("assets"),
            tools_home: agents_home.join("tools"),
            mcporter_home: home.join(".mcporter"),
            managed_state_home: home.join(MANAGED_STATE_SUBDIR),
            install_timeout,
            harnesses: vec![
                Harness::new(HarnessId::Claude, "claude", home.join(".claude"))
                    .with_instruction_file(CLAUDE_AGENT_FILE),
                Harness::new(HarnessId::Codex, "codex", home.join(".codex")),
                Harness::new(
                    HarnessId::Opencode,
                    "opencode",
                    home.join(".config").join("opencode"),
                ),
                Harness::new(HarnessId::Pi, "pi", home.join(".pi"))
                    .with_runtime_subdir("agent")
                    .with_compat_managed_entries(PI_COMPAT_MANAGED_ENTRIES),
                Harness::new(HarnessId::Omp, "omp", home.join(".omp")).with_runtime_subdir("agent"),
            ],
        }
    }

    pub(super) fn harness(&self, id: HarnessId) -> Option<&Harness> {
        self.harnesses.iter().find(|harness| harness.id == id)
    }
}

File: bin/harness.ts
import path from "node:path";

export const SOURCE_AGENT_FILE = "AGENTS.md";
const CLAUDE_AGENT_FILE = "CLAUDE.md";
const INSTALL_TIMEOUT_SECONDS = 120;
export const MANAGED_STATE_SUBDIR = ".local/share/agents/sync-managed";
const DEFAULT_PACKAGE_CACHE_SUBDIR = ".local/share/agents/pi-packages";

export type AssetRename = readonly [string, string];

const PI_COMPAT_MANAGED_ENTRIES = ["legacy"] as const;

export type HarnessHookSpec =
  | {
      readonly kind: "PackageBootstrap";
      readonly manifestFile?: string;
      readonly settingsFile?: string;
      readonly cacheSubdir?: string;
    }
  | {
      readonly kind: "ExtensionDeps";
      readonly rootDir: string;
    };

export type HarnessHook =
  | {
      readonly kind: "PackageBootstrap";
      readonly manifestFile: string;
      readonly settingsFile: string;
      readonly cacheSubdir: string;
    }
  | {
      readonly kind: "ExtensionDeps";
      readonly rootDir: string;
    };

export enum HarnessId {
  Claude = "Claude",
  Codex = "Codex",
  Opencode = "Opencode",
  Pi = "Pi",
  Omp = "Omp",
}

export interface HarnessSpec {
  readonly id: HarnessId;
  readonly sourceName: string;
  readonly home: string;
  readonly instructionFile?: string;
  readonly assetRenames?: readonly AssetRename[];
  readonly runtimeSubdir?: string;
  readonly compatManagedEntries?: readonly string[];
  readonly hooks?: readonly HarnessHookSpec[];
}

export class Harness {
  readonly id: HarnessId;
  readonly sourceName: string;
  readonly home: string;
  readonly instructionFile: string;
  readonly assetRenames: readonly AssetRename[];
  readonly runtimeSubdir: string | undefined;
  readonly compatManagedEntries: readonly string[];
  readonly hooks: readonly HarnessHook[];

  constructor(spec: HarnessSpec) {
    this.id = spec.id;
    this.sourceName = spec.sourceName;
    this.home = spec.home;
    this.instructionFile = spec.instructionFile ?? SOURCE_AGENT_FILE;
    this.assetRenames = spec.assetRenames ?? [];
    this.runtimeSubdir = spec.runtimeSubdir;
    this.compatManagedEntries = spec.compatManagedEntries ?? [];
    this.hooks = normalizeHooks(spec.hooks ?? []);
  }

  root(): string {
    return this.runtimeSubdir ? path.join(this.home, this.runtimeSubdir) : this.home;
  }

  sourceRoot(toolsHome: string): string {
    return this.runtimeSubdir
      ? path.join(toolsHome, this.sourceName, this.runtimeSubdir)
      : path.join(toolsHome, this.sourceName);
  }

  instructionFileName(): string {
    return this.instructionFile;
  }

  instructionTarget(): string {
    return path.join(this.root(), this.instructionFile);
  }

  renameAsset(assetName: string): string {
    const match = this.assetRenames.find(([src]) => src === assetName);
    return match ? match[1] : assetName;
  }

  managedStatePath(managedStateHome: string): string {
    return path.join(managedStateHome, `${this.sourceName}.json`);
  }
}

export class SyncEnv {
  readonly home: string;
  readonly assetsHome: string;
  readonly toolsHome: string;
  readonly mcporterHome: string;
  readonly managedStateHome: string;
  readonly installTimeoutMs: number;
  readonly harnesses: readonly Harness[];

  constructor(
    home: string,
    assetsHome: string,
    toolsHome: string,
    mcporterHome: string,
    managedStateHome: string,
    installTimeoutMs: number,
    harnesses: readonly Harness[]
  ) {
    this.home = home;
    this.assetsHome = assetsHome;
    this.toolsHome = toolsHome;
    this.mcporterHome = mcporterHome;
    this.managedStateHome = managedStateHome;
    this.installTimeoutMs = installTimeoutMs;
    this.harnesses = harnesses;
  }

  static fromSystem(): SyncEnv {
    const home = process.env.HOME;
    if (!home) {
      throw new Error("missing HOME");
    }
    return SyncEnv.fromHome(home, INSTALL_TIMEOUT_SECONDS * 1000);
  }

  static fromHome(home: string, installTimeoutMs: number): SyncEnv {
    const agentsHome = path.join(home, ".config", "agents");
    const harnessSpecs: HarnessSpec[] = [
      {
        id: HarnessId.Claude,
        sourceName: "claude",
        home: path.join(home, ".claude"),
        instructionFile: CLAUDE_AGENT_FILE,
      },
      {
        id: HarnessId.Codex,
        sourceName: "codex",
        home: path.join(home, ".codex"),
      },
      {
        id: HarnessId.Opencode,
        sourceName: "opencode",
        home: path.join(home, ".config", "opencode"),
      },
      {
        id: HarnessId.Pi,
        sourceName: "pi",
        home: path.join(home, ".pi"),
        runtimeSubdir: "agent",
        compatManagedEntries: PI_COMPAT_MANAGED_ENTRIES,
        hooks: [
          {
            kind: "PackageBootstrap",
            manifestFile: "packages.json",
            settingsFile: "settings.json",
            cacheSubdir: DEFAULT_PACKAGE_CACHE_SUBDIR,
          },
          {
            kind: "ExtensionDeps",
            rootDir: "extensions",
          },
        ],
      },
      {
        id: HarnessId.Omp,
        sourceName: "omp",
        home: path.join(home, ".omp"),
        runtimeSubdir: "agent",
      },
    ];
    return new SyncEnv(
      home,
      path.join(agentsHome, "assets"),
      path.join(agentsHome, "tools"),
      path.join(home, ".mcporter"),
      path.join(home, MANAGED_STATE_SUBDIR),
      installTimeoutMs,
      harnessSpecs.map((spec) => new Harness(spec))
    );
  }

  harness(id: HarnessId): Harness | undefined {
    return this.harnesses.find((harness) => harness.id === id);
  }
}

function normalizeHooks(hooks: readonly HarnessHookSpec[]): HarnessHook[] {
  return hooks.map((hook) => {
    switch (hook.kind) {
      case "PackageBootstrap":
        return {
          kind: hook.kind,
          manifestFile: hook.manifestFile ?? "packages.json",
          settingsFile: hook.settingsFile ?? "settings.json",
          cacheSubdir: hook.cacheSubdir ?? DEFAULT_PACKAGE_CACHE_SUBDIR,
        };
      case "ExtensionDeps":
        return {
          kind: hook.kind,
          rootDir: hook.rootDir,
        };
    }
  });
}
